src/components/name-tag.js

```diff
@@ -0,0 +1,377 @@
+import anime from "animejs";
+import MovingAverage from "moving-average";
+import { getThemeColor } from "../utils/theme";
+import qsTruthy from "../utils/qs_truthy";
+import { findAncestorWithComponent } from "../utils/scene-graph";
+import { NAMETAG_VISIBILITY_DISTANCE_DEFAULT } from "../react-components/preferences-screen";
+import { THREE } from "aframe";
+import { setMatrixWorld } from "../utils/three-utils";
+
+const DEBUG = qsTruthy("debug");
+const NAMETAG_BACKGROUND_PADDING = 0.05;
+const NAMETAG_STATUS_BORDER_PADDING = 0.035;
+const NAMETAG_MIN_WIDTH = 0.6;
+const NAMETAG_HEIGHT = 0.25;
+const NAMETAG_OFFSET = 0.2;
+const TYPING_ANIM_SPEED = 150;
+const DISPLAY_NAME_LENGTH = 18;
+
+const ANIM_CONFIG = {
+  duration: 400,
+  easing: "easeOutQuart",
+  elasticity: 400,
+  loop: 0,
+  round: false
+};
+
+AFRAME.registerComponent("name-tag", {
+  schema: {},
+  init() {
+    this.store = window.APP.store;
+    this.displayName = null;
+    this.identityName = null;
+    this.isTalking = false;
+    this.isTyping = false;
+    this.isOwner = false;
+    this.isRecording = false;
+    this.isHandRaised = false;
+    this.volumeAvg = new MovingAverage(128);
+    this.isNametagVisible = true;
+    this.size = new THREE.Vector3();
+    this.avatarBBAA = new THREE.Box3();
+    this.avatarBBAASize = new THREE.Vector3();
+    this.avatarBBAACenter = new THREE.Vector3();
+    this.nametagHeight = 0;
+    this.tmpNametagVisible = false;
+
+    this.onPresenceUpdated = this.onPresenceUpdated.bind(this);
+    this.onModelLoading = this.onModelLoading.bind(this);
+    this.onModelLoaded = this.onModelLoaded.bind(this);
+    this.onAnalyserVolumeUpdated = this.onAnalyserVolumeUpdated.bind(this);
+    this.onStateChanged = this.onStateChanged.bind(this);
+
+    this.avatarRig = document.getElementById("avatar-rig");
+    this.ikRootEl = findAncestorWithComponent(this.el, "ik-root");
+    this.playerInfo = this.el.components["player-info"];
+
+    this.nametagEl = this.el.querySelector(".nametag");
```

Comment:
I didn't check every one, but `nametagEl` for example we never use directly, we always get the `object3D` off of it, so we should just store a ref to that directly. Even when we do use both the element and object3D, we should probably prefer the object3D reference. I think I will add that to our best-practices.md

Reply:
Make sense. Updated.